### PR TITLE
HybridGrid proto.

### PR DESCRIPTION
**Pull request by macmason**

### CODE REVIEW

**Comment by macmason** [comment]
Related to issue #197

File: cartographer/mapping_3d/hybrid_grid_test.cc [hunk 1]
```diff
@@ -177,6 +178,120 @@ TEST(HybridGridTest, TestIteration) {
...
+  }
+  // 105 is the hand-verified number of times we added something.
+  ASSERT_EQ(105, grid_size);
+
+  proto = ToProto(hybrid_grid);
+  ASSERT_EQ(grid_size, proto.indices_size());
+  ASSERT_EQ(grid_size, proto.values_size());
+
+  int proto_index = 0;
+  for (const auto& entry : hybrid_grid) {
```
Comment (wohe) [inline cartographer/mapping_3d/hybrid_grid_test.cc]: This is very close to repeating the implementation and thus does not provide a lot of confidence in the correctness. Did you consider verifying that `HybridGrid(ToProto(original)) == original` as a test case? Maybe with (pseudo) randomly generated content.

Similar for the other test case. Combining the two into one then probably makes sense.

File: cartographer/mapping_3d/hybrid_grid_test.cc [hunk 2]
```diff
@@ -177,6 +178,120 @@ TEST(HybridGridTest, TestIteration) {
...
+    ASSERT_EQ(entry.second, proto.values(proto_index));
+    proto_index++;
+  }
+}
+
+namespace {
+
+struct EigenComparator {
+  bool operator()(const Eigen::Vector3i lhs, const Eigen::Vector3i& rhs) {
+    return (lhs.x() < rhs.x()) || (lhs.x() == rhs.x() && lhs.y() < rhs.y()) ||
```
Comment (wohe) [inline cartographer/mapping_3d/hybrid_grid_test.cc]: Use `<tuple>` and do
```
return std::forward_as_tuple(lhs.x(), lhs.y(), lhs.z()) <
    std::forward_as_tuple(rhs.x(), rhs.y(), rhs.z());
```
or so?

File: cartographer/transform/transform.cc [hunk 3]
```diff
@@ -99,6 +103,14 @@ proto::Vector3d ToProto(const Eigen::Vector3d& vector) {
   return proto;
 }
 
+proto::Vector3i ToProto(const Eigen::Vector3i& vector) {
+  proto::Vector3i result;
```
Comment (macmason) [inline cartographer/transform/transform.cc]: Done.

File: cartographer/mapping_3d/hybrid_grid_test.cc (hunk 1 above)
Comment (macmason) [inline cartographer/mapping_3d/hybrid_grid_test.cc]: Major rewrite here; PTAL.

**Comment by macmason** [comment]
PTAL.

**Pull request merged**